src/Team-Services-Bot.Api/Dialogs/ConnectDialog.cs

```diff
@@ -61,49 +63,124 @@ public Task StartAsync(IDialogContext context)
-            if (match.Success)
+        private async Task MessageReceivedAsync(IDialogContext context, IAwaitable<IMessageActivity> result)
+        {
+            if (context == null)
             {
-                var account = match.Groups[1].Value;
-                context.UserData.SetCurrentAccount(account);
+                throw new ArgumentNullException(nameof(context));
+            }
 
-                var reply = context.MakeMessage();
-                var profiles = context.UserData.GetProfiles();
-                var profile = profiles.FirstOrDefault(p => p.Accounts.Any(a => a.Equals(account, StringComparison.OrdinalIgnoreCase)));
+            if (result == null)
+            {
+                throw new ArgumentNullException(nameof(result));
+            }
 
-                if (profile != null)
+            var activity = await result;
+            var pin = context.UserData.GetPin();
+            var profile = context.UserData.GetCurrentProfile();
+            var profiles = context.UserData.GetProfiles();
+            var reply = context.MakeMessage();
+            var text = (activity.Text ?? string.Empty).ToLowerInvariant();
+
+            // Pin was previously created, waiting for it to return.
+            if (this.isPinActivated)
+            {
+                var isPinHandled = await this.HandlePin(context, activity, pin);
+                if (!isPinHandled)
                 {
-                    context.UserData.SetCurrentProfile(profile);
-                    reply.Text = string.Format(Labels.ConnectedTo, account);
+                    return;
                 }
-                else
+            }
+            else
+            {
+                // Wait for the normal expected input.
+                var match = Regex.Match(text, CommandMatchConnect);
+                if (match.Success)
                 {
-                    var button = new CardAction
-                    {
-                        Value = string.Format(
-                            UrlOAuth,
-                            this.appId,
-                            activity.ChannelId,
-                            activity.From.Id,
-                            Scope,
-                            this.authorizeUrl),
-                        Type = activity.IsTeamsChannel() ? "openUrl" : "signin",
-                        Title = Labels.AuthenticationRequired
-                    };
-
-                    var card = new SigninCard(Labels.PleaseLogin, new List<CardAction> { button });
-
-                    reply.Attachments.Add(card);
+                    this.account = match.Groups[1].Value;
+                    this.teamProject = match.Groups[2].Value;
                 }
+            }
+
+            // No Profiles, so we have to login.
+            if (!profiles.Any() || profile == null)
+            {
+                await this.Login(context, activity, reply);
+                return;
```

Comment:
It needs a discussion. I rather have it like this, then have nested if statements. Which I find hard to read.